File: tests/python/TestGeoCoordinate2d.py
"""
Unit Tests for GeoCoordinate2d
"""

from __future__ import print_function
import unittest
from quadkey import GeoCoordinate2d
from quadkey import Longitude
from quadkey import Latitude


class TestGeoCoordinate2d(unittest.TestCase):
    """
    Class for running unit tests.
    """

    def test_default_ctor(self):
        """
        Test default ctor
        """
        empty = GeoCoordinate2d()
        self.assertEquals(Longitude(0.0), empty.longitude())
        self.assertEquals(Latitude(0.0), empty.latitude())

    def test_ctor(self):
        """
        Test ctor
        """
        coord = GeoCoordinate2d(Longitude(0.0), Latitude(0.0))
        self.assertEquals(Longitude(0.0), coord.longitude())
        self.assertEquals(Latitude(0.0), coord.latitude())

        coord = GeoCoordinate2d(Longitude(-180.001), Latitude(-95.0))
        self.assertEquals(Longitude(-180.0), coord.longitude())
        self.assertEquals(Latitude(-90.0), coord.latitude())

        coord = GeoCoordinate2d(Longitude(-180.001), Latitude(95.0))
        self.assertEquals(Longitude(-180.0), coord.longitude())
        self.assertEquals(Latitude(90.0), coord.latitude())

        coord = GeoCoordinate2d(Longitude(180.001), Latitude(95.0))
        self.assertEquals(Longitude(180.0), coord.longitude())
        self.assertEquals(Latitude(90.0), coord.latitude())

        coord = GeoCoordinate2d(Longitude(180.001), Latitude(-95.0))
        self.assertEquals(Longitude(180.0), coord.longitude())
        self.assertEquals(Latitude(-90.0), coord.latitude())

    def test_near(self):
        """
        Test to see if one coord is near another
        """
        lon = Longitude(0.001)
        self.assertTrue(lon.near(Longitude(0.0009), 1e-3))

        lon = Longitude(0.001)
        self.assertFalse(lon.near(Longitude(0.0009), 1e-4))

        lat = Latitude(0.001)
        self.assertTrue(lat.near(Latitude(0.0009), 1e-3))

        lat = Latitude(0.001)
        self.assertFalse(lat.near(Latitude(0.0009), 1e-4))

        coord = GeoCoordinate2d(Longitude(0.001), Latitude(0.001))
        self.assertTrue(coord.near(GeoCoordinate2d(Longitude(0.0009),
                                                   Latitude(0.0009)), 1e-3))

        coord = GeoCoordinate2d(Longitude(0.001), Latitude(0.001))
        self.assertFalse(coord.near(GeoCoordinate2d(Longitude(0.001),
                                                    Latitude(0.0009)), 1e-4))

        coord = GeoCoordinate2d(Longitude(0.001), Latitude(0.001))
        self.assertFalse(coord.near(GeoCoordinate2d(Longitude(0.0009),
                                                    Latitude(0.001)), 1e-4))

    def test_operator_less_than(self):
        """
        Tests less than operator.
        """
        sw = GeoCoordinate2d(Longitude(-0.1), Latitude(-0.1))
        ne = GeoCoordinate2d(Longitude(+0.1), Latitude(+0.1))

        self.assertTrue(sw.longitude() < ne.longitude())
        self.assertTrue(sw.latitude() < ne.latitude())

    def test_operator_less_than_or_equal(self):
        """
        Tests less than or equal operator
        """
        sw = GeoCoordinate2d(Longitude(-0.1), Latitude(-0.1))
        ne = GeoCoordinate2d(Longitude(+0.1), Latitude(+0.1))
        self.assertTrue(sw.longitude() <= ne.longitude())
        self.assertTrue(sw.latitude() <= ne.latitude())

        sw = GeoCoordinate2d(Longitude(-0.1), Latitude(-0.1))
        ne = GeoCoordinate2d(Longitude(-0.1), Latitude(-0.1))
        self.assertTrue(sw.longitude() <= ne.longitude())
        self.assertTrue(sw.latitude() <= ne.latitude())

    def test_operator_equal(self):
        """
        Tests for equality of GeoCoordinate2d
        """
        sw = GeoCoordinate2d(Longitude(-0.1), Latitude(-0.1))
        ne = GeoCoordinate2d(Longitude(-0.1), Latitude(-0.1))
        self.assertEquals(sw.longitude(), ne.longitude())
        self.assertEquals(sw.latitude(), ne.latitude())


def main():
    """
    Bootstrap function for this module
    """
    suite = unittest.TestLoader().loadTestsFromTestCase(TestGeoCoordinate2d)
    unittest.TextTestRunner(verbosity=2).run(suite)

File: tests/test_geo_coordinate2d.cpp
#include <gtest/gtest.h>

#include "quad_key/geo_coordinate2d.h"

#include <iostream>

namespace quad_keys
{
  class test_geo_coordinate2d : public ::testing::Test
  {
  protected:
    virtual void SetUp() {}
    virtual void TearDown() {}
  };

  TEST(GEOCOORDINATE2D, default_ctor)
  {
    geo_coordinate2d empty;
    EXPECT_EQ(longitude(0.0), empty.get_longitude());
    EXPECT_EQ(latitude(0.0), empty.get_latitude());
  }

  TEST(GEOCOORDINATE2D, ctor)
  {
    {
      geo_coordinate2d coord(longitude(0.0), latitude(0.0));
      EXPECT_EQ(longitude(0.0), coord.get_longitude());
      EXPECT_EQ(latitude(0.0), coord.get_latitude());
    }

    {
      geo_coordinate2d coord(longitude(-180.001), latitude(-95.0));
      EXPECT_EQ(longitude(-180.0), coord.get_longitude());
      EXPECT_EQ(latitude(-90.0), coord.get_latitude());
    }

    {
      geo_coordinate2d coord(longitude(-180.001), latitude(95.0));
      EXPECT_EQ(longitude(-180.0), coord.get_longitude());
      EXPECT_EQ(latitude(90.0), coord.get_latitude());
    }

    {
      geo_coordinate2d coord(longitude(180.001), latitude(95.0));
      EXPECT_EQ(longitude(180.0), coord.get_longitude());
      EXPECT_EQ(latitude(90.0), coord.get_latitude());
    }

    {
      geo_coordinate2d coord(longitude(180.001), latitude(-95.0));
      EXPECT_EQ(longitude(180.0), coord.get_longitude());
      EXPECT_EQ(latitude(-90.0), coord.get_latitude());
    }
  }

  TEST(GEOCOORDINATE2D, near)
  {
    {
      longitude lon(0.001);
      EXPECT_TRUE(lon.near(longitude(0.0009), 1e-3));
    }

    {
      longitude lon(0.001);
      EXPECT_FALSE(lon.near(longitude(0.0009), 1e-4));
    }

    {
      latitude lat(0.001);
      EXPECT_TRUE(lat.near(latitude(0.0009), 1e-3));
    }

    {
      latitude lat(0.001);
      EXPECT_FALSE(lat.near(latitude(0.0009), 1e-4));
    }

    {
      geo_coordinate2d coord(longitude(0.001), latitude(0.001));
      EXPECT_TRUE(coord.near(geo_coordinate2d(longitude(0.0009),
        latitude(0.0009)), 1e-3));
    }

    {
      geo_coordinate2d coord(longitude(0.001), latitude(0.001));
      EXPECT_FALSE(coord.near(geo_coordinate2d(longitude(0.001),
        latitude(0.0009)), 1e-4));
    }

    {
      geo_coordinate2d coord(longitude(0.001), latitude(0.001));
      EXPECT_FALSE(coord.near(geo_coordinate2d(longitude(0.0009),
        latitude(0.001)), 1e-4));
    }
  }

  TEST(GEOCOORDINATE2D, operator_lt)
  {
    geo_coordinate2d sw(longitude(-0.1), latitude(-0.1));
    geo_coordinate2d ne(longitude(+0.1), latitude(+0.1));

    EXPECT_TRUE(sw.get_longitude() < ne.get_longitude());
    EXPECT_TRUE(sw.get_latitude() < ne.get_latitude());
  }

  TEST(GEOCOORDINATE2D, operator_lteq)
  {
    {
      geo_coordinate2d sw(longitude(-0.1), latitude(-0.1));
      geo_coordinate2d ne(longitude(+0.1), latitude(+0.1));

      EXPECT_TRUE(sw.get_longitude() <= ne.get_longitude());
      EXPECT_TRUE(sw.get_latitude() <= ne.get_latitude());
    }

    {
      geo_coordinate2d sw(longitude(-0.1), latitude(-0.1));
      geo_coordinate2d ne(longitude(-0.1), latitude(-0.1));

      EXPECT_TRUE(sw.get_longitude() <= ne.get_longitude());
      EXPECT_TRUE(sw.get_latitude() <= ne.get_latitude());
    }
  }

  TEST(GEOCOORDINATE2D, operator_eq)
  {
    geo_coordinate2d sw(longitude(-0.1), latitude(-0.1));
    geo_coordinate2d ne(longitude(-0.1), latitude(-0.1));

    EXPECT_TRUE(sw.get_longitude() == ne.get_longitude());
    EXPECT_TRUE(sw.get_latitude() == ne.get_latitude());
  }
}
